File: 5일차/그래프 이동.cpp
#include <stdio.h>
#include <queue>

using namespace std;

const int MAX = 21;

// N : 노드의 수, K : 경로의 수
int N, K;
int mat[MAX][MAX];	// 인접 배열
int visit[MAX];		// 방문에 대한 체크 : 1 -> 2, 2 -> 1 X

int main()
{
	scanf("%d %d", &N, &K);

	for (int i = 0; i < K; i++)
	{
		int to, from;
		scanf("%d %d", &from, &to);

		// from → to
		// to -> from
		mat[to][from] = 1;	// 1인경우 가는 경우
		mat[from][to] = 1;
	}

	// 1 번의 거리를 구한다.
	queue<int> q;
	q.push(1);

	visit[1] = 1;

	while (q.size())
	{
		int cur = q.front();
		q.pop();

		// 다음 탐색 경로를 검색합니다.
		for (int next = 1; next <= N; next++)
		{
			// cur -> next로 갈수 있고 && 방문을 안한 노드라면,
			if (mat[cur][next] == 1 && visit[next] == 0) {
				// 다음 거리 = 현재 거리 + 1
				visit[next] = visit[cur] + 1;	// 방문했다. (= 더 이상 방문하지마)
				q.push(next);
			}
		}
	}
}

File: 5일차/그래프 이동.py
from collections import deque

N, K = map(int, input().split())

# 인접 배열
mat = [[0] * (N+1) for i in range(N + 1)]

# 방문 정보
visit = []
for i in range(N + 1):
    visit.append(0)

for i in range(K):
    a, b = map(int, input().split())
    mat[b][a] = mat[a][b] = 1

# BFS
q = deque()
q.append(1)
visit[1] = 1

while len(q):
    cur = q.popleft()

    # 다음 이동경로 탐색
    for next in range(N + 1):
        # 현재에서 다음으로 이동 가능한 경로 이고, 방문한적이 없으면
        if mat[cur][next] == 1 and visit[next] == 0:
            # 거리를 추가로 더해주고, 다음 거리 = 현재 거리 + 1
            visit[next] = visit[cur] + 1
            q.append(next)

# 전체 거리 출력
for i in range (1, N + 1):
    print(visit[i] - 1)
